python/tank/util/shotgun.py

```diff
@@ -312,25 +285,80 @@ def create_sg_connection(user="default"):
     :returns: SG API instance
     """
-    api_handle, _ = __create_sg_connection(__get_sg_config(), evaluate_script_user=False, user=user)
+    # get connection parameters
+    config_path = __get_sg_config()
+    config_data = __get_sg_config_data(config_path, user)
+    api_handle = __create_sg_connection(config_data)
     return api_handle
 
+g_app_store_connection = None
 def create_sg_app_store_connection():
     """
-    Creates a shotgun connection to the tank app store.
+    Creates a shotgun connection that can be used to access the toolkit app store.
 
-    returns a tuple: (api_handle, script_user_entity)
-
-    The second part of the tuple represents the
-    user that was used to connect to the app store,
-    as a standard sg entity dictionary.
+    :returns: (sg, dict) where the first item is the shotgun api instance and the second 
+              is an sg entity dictionary (keys type/id) corresponding to to the user used
+              to connect to the app store.
     """
     global g_app_store_connection
     
-    if g_app_store_connection is None:
-        g_app_store_connection = __create_sg_connection(__get_app_store_config(), evaluate_script_user=True)
+    # maintain a cache for performance
+    if g_app_store_connection:
+        return g_app_store_connection
+    
+    # get app store credentials from shotgun
+    client_site_sg = get_sg_connection()
+    (script_name, script_key) = __get_app_store_key_from_shotgun(client_site_sg)
+    
+    # connect to the app store
+    config_data = {}
+    config_data["host"] = client_site_sg.base_url
+    config_data["api_script"] = script_name
+    config_data["api_key"] = script_key
+    config_data["http_proxy"] = client_site_sg.config.raw_http_proxy
+    app_store_sg = __create_sg_connection(config_data)
     
-    return g_app_store_connection
+    # now figure out the entity id for the script user    
+    script_user = app_store_sg.find_one("ApiUser", [["firstname", "is", script_name]], fields=["type", "id"])
+    if script_user is None:
+        raise TankError("Could not evaluate the current App Store User! Please contact support.")
+    
+    return (app_store_sg, script_user)
+
+
+def __get_app_store_key_from_shotgun(sg_connection):
+    """
+    Given a Shotgun url and script credentials, fetch the app store key
+    for this shotgun instance using a special controller method.
+    Returns a tuple with (app_store_script_name, app_store_auth_key)
+    
+    :param sg_connection: SG connection to the client site for which
+                          app store credentials should be retrieved.
+    :returns: tuple of strings with contents (script_name, script_key)
+    """
+    # handle proxy setup by pulling the proxy details from the main shotgun connection
+    if sg_connection.config.proxy_server:
+        
+        if sg_connection.config.proxy_user and sg_connection.config.proxy_pass:
```

Comment:
I just found this: https://github.com/shotgunsoftware/python-api/blob/master/shotgun_api3/shotgun.py#L1492
Maybe we could find a way to share this piece of code. This is pretty tricky code that shouldn't be retyped twice.